File: ACTIVIDADES/Actividad N°5/bubblesort.py
import random  # Para generar arreglos aleatorios
import time    # Para medir el tiempo de ejecución

# Función de ordenamiento Bubble Sort
def bubble_sort(arr):
    n = len(arr)  # Longitud del arreglo
    for i in range(n):  # Recorrido de n veces
        for j in range(n - i - 1):  # Comparar elementos adyacentes
            if arr[j] > arr[j + 1]:  # Si están desordenados
                arr[j], arr[j + 1] = arr[j + 1], arr[j]  # Intercambiarlos
        if i < 5:  # Solo mostrar las primeras 5 iteraciones
            print(f"Iteración {i + 1}: {arr[:10]}...")  # Mostrar primeros 10 elementos

# Función para medir tiempo y mostrar resultados
def probar_bubble_sort(nombre, arreglo):
    print(f"\n--- {nombre.upper()} ---")  # Título
    copia = arreglo.copy()  # Copiar arreglo para no modificar el original
    inicio = time.time()  # Tiempo antes de ordenar
    bubble_sort(copia)  # Ejecutar el algoritmo
    fin = time.time()  # Tiempo después de ordenar
    print(f"{nombre} - Tiempo: {fin - inicio:.6f} segundos")  # Mostrar tiempo

# Crear tres tipos de arreglos
n = 1000
aleatorio = random.sample(range(n), n)  # Aleatorio sin repetición
ordenado = list(range(n))  # Arreglo ya ordenado
inverso = list(range(n, 0, -1))  # Arreglo en orden descendente

# Ejecutar pruebas
print("Bubble Sort:")
probar_bubble_sort("Aleatorio", aleatorio)
probar_bubble_sort("Ordenado", ordenado)
probar_bubble_sort("Inverso", inverso)

File: ACTIVIDADES/Actividad N°5/bubblesort.cpp
#include <iostream>     // Para imprimir en consola
#include <vector>       // Para usar std::vector
#include <algorithm>    // Para std::swap
#include <ctime>        // Para medir tiempo de ejecución
#include <cstdlib>      // Para usar rand()

/*
Descripción:
Bubble Sort: Algoritmo de ordenamiento simple que compara e intercambia elementos adyacentes,
burbujeando el mayor al final en cada pasada. Es ineficiente para listas grandes.
*/

// Función que ordena un vector usando Bubble Sort
void bubbleSort(std::vector<int>& arr) {
    clock_t inicio = clock();  // Marca el inicio del tiempo

    int n = arr.size();  // Obtener el tamaño del vector
    for (int i = 0; i < n - 1; ++i) {  // Repetir n-1 veces
        for (std::vector<int>::size_type j = 0; j < n - i - 1; ++j) {  // Usar el tipo correcto
            if (arr[j] > arr[j + 1]) {  // Si están fuera de orden
                std::swap(arr[j], arr[j + 1]);  // Intercambiar
            }
        }

        // Mostrar el estado del vector después de cada pasada, solo los primeros 10 elementos
        if (i < 5) {  // Solo mostrar las primeras 5 pasadas
            std::cout << "Pasada " << i + 1 << ": ";
            for (std::vector<int>::size_type j = 0; j < 10 && j < arr.size(); ++j) {
                std::cout << arr[j] << " ";  // Mostrar primeros 10 elementos
            }
            std::cout << "..." << std::endl;
        }
    }

    clock_t fin = clock();  // Marca el final del tiempo
    double tiempo = double(fin - inicio) / CLOCKS_PER_SEC;  // Calcular tiempo
    std::cout << "Tiempo total de ejecucion: " << tiempo << " segundos" << std::endl;
}

// Función para generar un arreglo aleatorio de tamaño n
std::vector<int> generarArregloAleatorio(int n) {
    std::vector<int> arr(n);
    for (int i = 0; i < n; ++i) {
        arr[i] = rand() % 1000;  // Llenar con números aleatorios entre 0 y 999
    }
    return arr;
}

// Función principal
int main() {
    srand(time(0));  // Inicializar la semilla aleatoria

    int n = 1000;  // Tamaño del arreglo

    // Crear tres vectores de ejemplo
    std::vector<int> aleatorio = generarArregloAleatorio(n);
    std::vector<int> ordenado(n);
    for (int i = 0; i < n; ++i) {
        ordenado[i] = i + 1;  // Arreglo ya ordenado
    }
    std::vector<int> inverso(n);
    for (int i = 0; i < n; ++i) {
        inverso[i] = n - i;  // Arreglo en orden descendente
    }

    // Ejecutar Bubble Sort para cada tipo de vector
    std::cout << "--- ALEATORIO ---" << std::endl;
    bubbleSort(aleatorio);

    std::cout << "\n--- ORDENADO ---" << std::endl;
    bubbleSort(ordenado);

    std::cout << "\n--- INVERSO ---" << std::endl;
    bubbleSort(inverso);

    return 0;
}
